File: benchmarks/reference-library/spike.cpp
// The arithmetic half of the reference-library spike in Boost's cpp_bin_float,
// the counterpart of spike.swift; README.md explains, reproduce.py builds it.

#define BOOST_MP_STANDALONE
#include <boost/multiprecision/cpp_bin_float.hpp>
#include <chrono>
#include <iostream>
#include <cmath>
template<unsigned Bits> void run(int digits) {
 using F=boost::multiprecision::number<boost::multiprecision::cpp_bin_float<Bits,boost::multiprecision::digit_base_2>>;
 F cr=F("-0.743643887037151"),ci=F("0.13182590390533"); double checksum=0;
 std::cout<<"{\"library\":\"Boost-1.90-cpp_bin_float\",\"digits\":"<<digits<<",\"bits\":"<<Bits<<",\"seconds\":[";
 for(int r=0;r<5;r++) {
  auto start=std::chrono::steady_clock::now();
  for(int repeat=0;repeat<10;repeat++) {
   F x=0,y=0;
   for(int i=0;i<1000;i++){F xx=x*x-y*y; y=2*x*y+ci;x=xx+cr;}
   checksum+=x.template convert_to<double>();
  }
  if(r)std::cout<<",";
  std::cout<<std::chrono::duration<double>(std::chrono::steady_clock::now()-start).count();
 }
 std::cout<<"],\"checksum\":"<<checksum<<"}\n";
}
int main(){run<397>(100);run<3386>(1000);}

File: benchmarks/reference-library/spike.swift
// The arithmetic half of the reference-library spike: ten 1,000-step orbits in
// the vendored BigInt's fixed point, at 100 and 1,000 digits.  spike.cpp does
// the same in Boost; README.md explains, reproduce.py builds and runs both.

import Foundation

@main struct Spike {
  static func main() {
    for digits in [100, 1000] {
      let bits = Int(ceil(Double(digits) * log2(10))) + 64
      let unit = BigInt(1) << bits
      let cr = -unit * 743_643_887_037_151 / 1_000_000_000_000_000
      let ci = unit * 13_182_590_390_533 / 100_000_000_000_000
      var times: [Double] = []
      var checksum = 0.0
      for _ in 0..<5 {
        let start = ProcessInfo.processInfo.systemUptime
        // Restart before escape: identical 1,000-step bounded orbit, 10 times.
        for _ in 0..<10 {
          var x = BigInt(0)
          var y = BigInt(0)
          for _ in 0..<1000 {
            let xx = (x * x - y * y) >> bits
            y = ((2 * x * y) >> bits) + ci
            x = xx + cr
          }
          checksum += Double(x >> max(0, bits - 40)) / pow(2, 40)
        }
        times.append(ProcessInfo.processInfo.systemUptime - start)
      }
      print(
        "{\"library\":\"BigInt-5.7.0-fixed\",\"digits\":\(digits),\"bits\":\(bits),\"seconds\":\(times),\"checksum\":\(checksum)}"
      )
    }
  }
}
